cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java

```diff
@@ -219,41 +223,33 @@ public void addInput(Input input, @Nullable Class<?> mapperCls) {
     }
   }
 
-  private void addOutput(String alias, OutputFormatProvider outputFormatProvider) {
+  @Override
+  public void addOutput(Output output) {
+    String alias = output.getAlias();
     if (this.outputs.containsKey(alias)) {
       throw new IllegalArgumentException("Output already configured: " + alias);
     }
-    this.outputs.put(alias, new ProvidedOutput(alias, outputFormatProvider));
-  }
 
-  @Override
-  public void addOutput(Output output) {
+    ProvidedOutput providedOutput;
     if (output instanceof Output.DatasetOutput) {
-      Output.DatasetOutput datasetOutput = ((Output.DatasetOutput) output);
-      String datasetNamespace = datasetOutput.getNamespace();
-      if (datasetNamespace == null) {
-        datasetNamespace = getNamespace();
-      }
-      String datasetName = output.getName();
-      Map<String, String> arguments = ((Output.DatasetOutput) output).getArguments();
-
-      // we can delay the instantiation of the Dataset to later, but for now, we still have to maintain backwards
-      // compatibility for the #setOutput(String, Dataset) method, so delaying the instantiation of this dataset will
-      // bring about code complexity without much benefit. Once #setOutput(String, Dataset) is removed, we can postpone
-      // this dataset instantiation
-      DatasetOutputFormatProvider outputFormatProvider =
-        new DatasetOutputFormatProvider(datasetNamespace, datasetName, arguments,
-                                        getDataset(datasetNamespace, datasetName, arguments, AccessType.WRITE),
-                                        MapReduceBatchWritableOutputFormat.class);
-      addOutput(output.getAlias(), outputFormatProvider);
-
+      providedOutput = Outputs.transform((Output.DatasetOutput) output, this);
     } else if (output instanceof Output.OutputFormatProviderOutput) {
-      addOutput(output.getAlias(), ((Output.OutputFormatProviderOutput) output).getOutputFormatProvider());
+      OutputFormatProvider outputFormatProvider =
+        ((Output.OutputFormatProviderOutput) output).getOutputFormatProvider();
+      if (outputFormatProvider instanceof DatasetOutputCommitter) {
```

Comment:
This disallowing is not backwards-compatible. See the discussion here: https://github.com/caskdata/cdap/pull/9442#discussion_r134637657